sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergUtils.java

```diff
@@ -99,36 +101,128 @@ public static Schema icebergSchemaToBeamSchema(final org.apache.iceberg.Schema s
     return builder.build();
   }
 
-  public static Schema icebergStructTypeToBeamSchema(final Types.StructType struct) {
+  private static Schema icebergStructTypeToBeamSchema(final Types.StructType struct) {
     Schema.Builder builder = Schema.builder();
     for (Types.NestedField f : struct.fields()) {
       builder.addField(icebergFieldToBeamField(f));
     }
     return builder.build();
   }
 
-  public static Types.NestedField beamFieldToIcebergField(int fieldId, final Schema.Field field) {
-    @Nullable Type icebergType = BEAM_TYPES_TO_ICEBERG_TYPES.get(field.getType());
+  /**
+   * Represents an Object (in practice, either {@link Type} or {@link Types.NestedField}) along with
+   * the most recent (max) ID that has been used to build this object.
+   *
+   * <p>Iceberg Schema fields are required to have unique IDs. This includes unique IDs for a {@link
+   * Types.ListType}'s collection type, a {@link Types.MapType}'s key type and value type, and
+   * nested {@link Types.StructType}s. When constructing any of these types, we use multiple unique
+   * ID's for the type's components. The {@code maxId} in this object represents the most recent ID
+   * used after building this type. This helps signal that the next field we construct should have
+   * an ID greater than this one.
+   */
+  private static class ObjectAndMaxId<T> {
+    int maxId;
+    T object;
 
-    if (icebergType != null) {
-      return Types.NestedField.of(
-          fieldId, field.getType().getNullable(), field.getName(), icebergType);
-    } else {
-      return Types.NestedField.of(
-          fieldId, field.getType().getNullable(), field.getName(), Types.StringType.get());
+    ObjectAndMaxId(int id, T object) {
+      this.maxId = id;
+      this.object = object;
     }
   }
 
+  private static ObjectAndMaxId<Type> beamFieldTypeToIcebergFieldType(
+      int fieldId, Schema.FieldType beamType) {
+    if (BEAM_TYPES_TO_ICEBERG_TYPES.containsKey(beamType.getTypeName())) {
+      return new ObjectAndMaxId<>(fieldId, BEAM_TYPES_TO_ICEBERG_TYPES.get(beamType.getTypeName()));
+    } else if (beamType.getTypeName().isCollectionType()) { // ARRAY or ITERABLE
+      // List ID needs to be unique from the NestedField that contains this ListType
+      int listId = fieldId + 1;
```

Comment:
is this because iceberg listId index starts from 1? Good to have a comment here

this conversion is seen elsewhere, maybe put `int listId = fieldId + 1;` at the beginning of this method

0-indexed and 1-indexed systems conversion is a well known annoying and error-prone thing

Reply:
Tried to explain this here: https://github.com/apache/beam/pull/31958/files/#diff-88fae921b5fb1c41f0f30e9002fe3c165bcd4e244491b3f092fe0b9be5778111R112-R123

Each field needs to have a unique global number within the scope of a Schema object. You can see this in our tests for example: https://github.com/apache/beam/pull/31958/files/#diff-62fb088ee662ceb6f351514aadf6bfc616a16e3afce17f52b435cb9a4a47569eR372-R381

Reply:
For example if the parent field ID is "5", then we increment and give a unique ID "6" to this List type. 

Not all field types require an ID (e.g. Primitive types). But some do, for example a Map type requires an ID for its key type and an ID for its value type (again, all have to be unique). Another example is Struct type which naturally requires an ID for each of its nested field(s).